compaction/compaction_manager.cc

```diff
@@ -1524,23 +1524,27 @@ class cleanup_sstables_compaction_task_executor : public compaction_task_executo
         co_return std::nullopt;
     }
 private:
-    // Releases reference to cleaned files such that respective used disk space can be freed.
-    void release_exhausted(std::vector<sstables::shared_sstable> exhausted_sstables) {
-        _compacting.release_compacting(exhausted_sstables);
-    }
-
     future<> run_cleanup_job(sstables::compaction_descriptor descriptor) {
         co_await coroutine::switch_to(_cm.compaction_sg().cpu);
 
+        // Releases reference to cleaned files such that respective used disk space can be freed.
+        auto release_exhausted = [this, &descriptor] (std::vector<sstables::shared_sstable> exhausted_sstables) mutable {
+            auto exhausted = boost::copy_range<std::unordered_set<sstables::shared_sstable>>(exhausted_sstables);
+            auto e = boost::range::remove_if(descriptor.sstables, [&] (const sstables::shared_sstable& sst) {
+                return exhausted.contains(sst);
+            });
+            descriptor.sstables.erase(e, descriptor.sstables.end());
+            _compacting.release_compacting(exhausted_sstables);
+        };
```

Comment:
I think this is incomplete. What about sstables that are being compacted in regular or major compaction? Or are sstables that participate in cleanup excluded from regular compaction?

Reply:
@avikivity sstables intended for cleanup are excluded from regular compaction (and therefore they require `_compacting.release_compacting(exhausted_sstables);` above.
We changed to cleanup to be equivalent to regular compaction by using the same bucket selection logic.

BUT, one issue I still think we might have there with ICS (that is orthogonal to this fix), is that we pick particular sstables for cleanup based on the token ownership heuristic, meaning we're breaking sstable runs.
We should consider picking complete runs for cleanup instead and when compacting them, apply token filtering only for the fragments that require cleanup (based on https://github.com/scylladb/scylladb/pull/12422)

Reply:
I think Avi thought we're trying to exclude compacting sstables, but what we're doing here is to exclude sstables which were already compacted, in order to allow their space to be reclaimed.

Reply:
Benny, your observation is correct, but your conclusion may not.

SSTable runs are not really broken. 

Cleanup on ICS will extract unowned data from existing run fragments, and produce a single new run for each level containing owned data.

So existing runs have holes, whereas new ones don't. Regular compaction logic kicks in later, in order to compact similar-sized runs together.

Existing run and new one at a given level L won't even contain the same token T, meaning that read can still open a single fragment in a level L to satisfy a single-key query. And transient accumulation of runs in a given level L is perfectly fine with size tiered strategies.

So unless we have a good reason to, I don't think we need to do anything.

Reply:
By "broken" I didn't mean there's a bug, but rather that existing runs might get broken into 2 pieces:
1. The fragments that don't require cleanup will remain in the original runs and will be subject to regular compaction.
2. Fragments that are marked for cleanup will get cleaned up by cleanup compaction into new (smaller runs).

Like you said, I'm not sure if that's worth fixing.

Reply:
> By "broken" I didn't mean there's a bug, but rather that existing runs might get broken into 2 pieces:
> 1. The fragments that don't require cleanup will remain in the original runs and will be subject to regular compaction.
> 2. Fragments that are marked for cleanup will get cleaned up by cleanup compaction into new (smaller runs).
> 
> Like you said, I'm not sure if that's worth fixing.

Exactly. It's a very interesting observation though. Thanks for bringing it up